File: 1.30.1.js
// procedures as arguments
function sum(term, a, next, b) {
  if (a > b) {
    return 0;
  } else {
    return term(a) + sum(term, next(a), next, b);
  }
}

function identity(x) {
  return x;
}

function inc(x) {
  return x + 1;
}

function sum_integers_01(a,b) {
  return sum(identity, a, inc, b);
}

function sum_integers_02(a,b) {
  return sum(function(a) { return a; }, a,
             function(a) { return a + 1; }, b)
}

console.log(sum_integers_01(1,10));
console.log(sum_integers_02(1,10));

console.log(sum_integers_01(1,10) === sum_integers_01(1,10));

File: 1.30.1.py
def printdocstr(fn):
    def wrap(*args, **kwargs):
        print(fn.__doc__)
        return fn(fn, *args, **kwargs)
    return wrap

def sum_01(term, a, nxt, b):
    if (a > b):
        return 0
    else:
        return term(a) + sum_01(term, nxt(a), nxt, b)

def identity(x):
    return x;

def inc(x):
    return x + 1;

@printdocstr
def si_01(fn,a,b):
    """Just the SICP sample"""
    return sum_01(identity, a, inc, b)

print(si_01(1,10))
